bayes_kit/ensemble.py

```diff
@@ -1,68 +1,91 @@
-from typing import Callable, Optional, Tuple
+from typing import Callable, Iterator, Optional, Tuple
 from numpy.typing import NDArray
 import numpy as np
 
 from .model_types import LogDensityModel
 
+Sample = NDArray[np.float64]
 
-class Stretcher:
+class AffineInvariantWalker:
     """
+    An implementation of the affine-invariant ensemble sampler of
+    Goodman and Weare (2010).  
+
+    References:
     Goodman, J. and Weare, J., 2010. Ensemble samplers with affine invariance.
     *Communications in Applied Mathematics and Computational Science*
     5(1):65--80.
     """
 
-    # def __init__(
-    #         self,
-    #         model: LogDensityModel,
-    #         a: Optional[float] = None,
-    #         walkers: Optional[int] = None
-    #         init: Optional[NDarray[np.float64]] = None)
-    #     ):
-    #     self._model = model
-    #     self._dim = self._model.dims()
-    #     if a != None and a < 1:
-    #         raise ValueError(f"stretch bound must be greater than or equal to 1; found {a=}")
-    #     self._a = a
-    #     self._sqrt_a = np.sqrt(a)
-    #     self._inv_sqrt_a = 1 / self._sqrt_a
-    #     if walkers != NONE and (walkers <= 0 or walkers % 2 != 0) :
-    #         raise ValueError(f"walkers must be strictly positive, even integer; found {walkers=}")
-    #     self._walkers = walkers or 2 * self._dim
-    #     self._halfwalkers = a / 2
-    #     self._drawshape = (self._walkers, self._dim)
-    #     if init != None and init.shape != self._drawshape:
-    #         raise ValueError(f"init must be shape of draw {self._drawshape}; found {init.shape=}")
-    #     self._thetas = init or np.random.normal(size=self._drawshape)
-    #     self._firsthalf = range(halfwalkers)
-    #     self._secondhalf = range(halfwalkers, walkers)
+    def __init__(
+            self,
+            model: LogDensityModel,
+            a: Optional[float] = None,
+            walkers: Optional[int] = None,
+            init: Optional[NDArray[np.float64]] = None
+        ):
+        """
+        Initialize the sampler with a log density model, and optionally
```

Comment:
We haven't previously been writing full docstrings, so if we start we should settle on a format. I tend to prefer the [numpy style](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) ([example](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy)), but anything is fine by me as long as it is consistent and it is an option that Sphinx can read down the line.

(aside: I use [this VSCode extension](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring) to generate blank docstrings in my chosen format)

Reply:
Just to note this somewhere written: Bob and I discussed in person and decided to go with the [Google style](https://google.github.io/styleguide/pyguide.html#381-docstrings) ([example](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google)), mostly because it uses a lot less vertical space than the numpy style.

Reply:
I added a CONTRIBUTING.md file with this PR outlining some of our dev practices.